sklearn/metrics/_classification.py

```diff
@@ -337,62 +337,76 @@ def confusion_matrix(
     (np.int64(0), np.int64(2), np.int64(1), np.int64(1))
     """
     y_true, y_pred = attach_unique(y_true, y_pred)
+    xp, _, device_ = get_namespace_and_device(y_true, y_pred, labels, sample_weight)
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     if y_type not in ("binary", "multiclass"):
         raise ValueError("%s is not supported" % y_type)
 
     if labels is None:
         labels = unique_labels(y_true, y_pred)
     else:
-        labels = np.asarray(labels)
+        labels = xp.asarray(labels)
         n_labels = labels.size
         if n_labels == 0:
             raise ValueError("'labels' should contains at least one label.")
         elif y_true.size == 0:
-            return np.zeros((n_labels, n_labels), dtype=int)
-        elif len(np.intersect1d(y_true, labels)) == 0:
+            return xp.zeros((n_labels, n_labels), dtype=xp.int64, device=device_)
+        elif not xp.isin(labels, y_true).any():
```

Comment:
This line is currently only tested for ndarrays.

`xp.isin()` is not existing in `array_api_strict` and I am trying to find an alternative way that also works in the strict definition.

Reply:
Can you use `sklearn.utils._array_api._isin`? Looking at `isin` status in array API I found https://github.com/data-apis/array-api/issues/854 that mentions scikit-learn has an implementation of it.

Reply:
Oh yes, I can! :sparkles: 
Thanks for the suggestion.

Reply:
I guess the fact that it was not failing with `xp.isin` makes me think that some additional tests would be needed to test this part of the code with array API (non numpy) inputs

Reply:
I agree. All the branches down the control flow should be tested (maybe with some exceptions like just raising error messages).

Reply:
I will add a test.